wallyworld/juju
# Wire up new cmr macaroon authenticator

## Description of change

Hook the new cross model offer authenticator into the api server. This provides a http endpoint for discharging the "has-offer-permission" third party caveat. Now, macaroons used to validate access to ffers are given a 2 minute expiry time. When they expire, a discharge error pointing to the new endpoint is generated.

The crossmodel api facade methods all now have the capability to transparently recognise the discharge error and perform the discharge. The workers which use the facade are none  the wiser. The facade methods essentially try the facade call up to 2 times - once initially, and then a second time with discharge macaroon if the first time failed with a discharge error and the third party caveats could be successfully discharged. To make this all possible, the bakery client off the api connection is exposed so it can be used to do the discharge in the api client.

## QA steps

Run up a CMR scenario and check logs for errors.
Everything should work as previously with the old auth.
